read-fonts/src/tables/gsub/closure.rs

```diff
@@ -28,214 +27,260 @@ use crate::tables::layout::{
+            {
+                let (count, covered) = self
+                    .done_lookups_glyphs
+                    .entry(lookup_index)
+                    .or_insert((0, IntSet::empty()));
+
+                if *count != self.glyphs.len() {
+                    *count = self.glyphs.len();
+                    covered.clear();
+                }
             }
-            covered
-                .get_or_insert_with(Default::default)
-                .extend(current_glyphs.unwrap_or(self.glyphs).iter());
-            true
+
+            let mut cur_glyphs = IntSet::empty();
+            {
+                let covered = &self.done_lookups_glyphs.get(&lookup_index).unwrap().1;
+                //TODO: add IntSet::is_subset
+                if self
+                    .parent_active_glyphs()
+                    .iter()
+                    .all(|g| covered.contains(g))
+                {
+                    return true;
+                }
+                cur_glyphs.extend(self.parent_active_glyphs().iter());
+            }
+
+            let (_, covered) = self.done_lookups_glyphs.get_mut(&lookup_index).unwrap();
+            covered.union(&cur_glyphs);
+            false
+        }
+
+        pub(super) fn glyphs(&self) -> &IntSet<GlyphId> {
+            self.glyphs
+        }
+
+        pub(super) fn add(&mut self, gid: GlyphId) {
+            self.output.insert(gid);
+        }
+
+        pub(super) fn add_glyphs(&mut self, iter: impl IntoIterator<Item = GlyphId>) {
+            self.output.extend(iter)
+        }
+
+        pub(super) fn flush(&mut self) {
+            self.glyphs.union(&self.output);
+            self.output.clear();
+            self.active_glyphs_stack.clear();
         }
     }
 }
 
 use ctx::ClosureCtx;
 
-/// a lookup referenced by a contextual lookup
-#[derive(Debug)]
-struct ContextualLookupRef {
-    lookup_id: u16,
-    // 'none' means the graph is too complex, assume all glyphs are active
-    active_glyphs: Option<IntSet<GlyphId16>>,
-}
-
 /// A trait for tables which participate in closure
 trait GlyphClosure {
     /// Update the set of glyphs with any glyphs reachable via substitution.
-    fn add_reachable_glyphs(&self, ctx: &mut ClosureCtx) -> Result<(), ReadError>;
+    fn closure_glyphs(
+        &self,
+        ctx: &mut ClosureCtx,
+        lookup_list: &SubstitutionLookupList,
+        lookup_index: u16,
+    ) -> Result<(), ReadError>;
+
+    fn may_have_non_1to1(&self) -> Result<bool, ReadError> {
+        Ok(false)
+    }
 }
 
+const CLOSURE_MAX_STAGES: u8 = 12;
 impl Gsub<'_> {
     /// Return the set of glyphs reachable from the input set via any substitution.
     pub fn closure_glyphs(
         &self,
-        mut glyphs: IntSet<GlyphId16>,
-    ) -> Result<IntSet<GlyphId16>, ReadError> {
-        // we need to do this iteratively, since any glyph found in one pass
-        // over the lookups could also be the target of substitutions.
+        lookups: &IntSet<u16>,
```

Comment:
is this all based on  the harfbuzz? if so it would be useful to add some links to the original implementation source...